### PR TITLE
Torus passing through three points

### PR DESCRIPTION
**Pull request by stla**

### CODE REVIEW

File: src/primitives/torus.jl [hunk 1]
```diff
@@ -22,6 +22,21 @@ Torus(center::Point{3,T}, normal::Vec{3,T}, major, minor) where {T} =
...
+minor radius `minor`.
+"""
+function Torus(p1::Point{3}, p2::Point{3}, p3::Point{3}, minor)
+  c = Circle(p1, p2, p3)
+  O = center(c)
+  major = radius(c)
+  n⃗ = normal(Plane(p1, p2, p3))
+  T = typeof(major)
+  Torus(O, Vec{3,T}(n⃗), major, T(minor))
+end
```
Comment (juliohm) [inline src/primitives/torus.jl]:
```suggestion
function Torus(p1::Point{3,T}, p2::Point{3,T}, p3::Point{3,T}, minor) where {T}
  c = Circle(p1, p2, p3)
  p = Plane(p1, p2, p3)
  Torus(center(c), normal(p), radius(c), T(minor))
end
```

File: src/primitives/torus.jl [hunk 2]
```diff
@@ -22,6 +22,18 @@ Torus(center::Point{3,T}, normal::Vec{3,T}, major, minor) where {T} =
 Torus(center::Tuple, normal::Tuple, major, minor) = 
   Torus(Point(center), Vec(normal), major, minor)
 
+"""
+    Torus(p1, p2, p3)
```
Comment (juliohm) [inline src/primitives/torus.jl]:
```suggestion
    Torus(p1, p2, p3, minor)
```

**Comment by juliohm** [comment]
@stla I find this Torus constructor a little exotic. Is it common to construct Torus like this?

**Comment by stla** [comment]
> @stla I find this Torus constructor a little exotic. Is it common to construct Torus like this?

Well this is like a circle passing through three points with a "width" in addition. See [my repo here](https://github.com/stla/PyTorusThreePoints) for some "applications".

**Comment by stla** [comment]
also [this one](https://github.com/stla/villarceau)

File: src/primitives/torus.jl [hunk 3]
```diff
@@ -22,6 +22,18 @@ Torus(center::Point{3,T}, normal::Vec{3,T}, major, minor) where {T} =
...
+
+The torus whose equator passes through points `p1`, `p2` and `p3` and with
+minor radius `minor`.
+"""
+function Torus(p1::Point{3,T}, p2::Point{3,T}, p3::Point{3,T}, minor) where {T}
+  c = Circle(p1, p2, p3)
+  p = Plane(p1, p2, p3)
+  Torus(center(c), normal(p), radius(c), T(minor))
+end
+
```
Comment (juliohm) [inline src/primitives/torus.jl:36]:
```suggestion
function Torus(p1::Point{3,T}, p2::Point{3,T}, p3::Point{3,T}, minor) where {T}
  c = Circle(p1, p2, p3)
  p = Plane(p1, p2, p3)
  Torus(center(c), normal(p), radius(c), T(minor))
end

Torus(p1::Tuple, p2::Tuple, p3::Tuple, minor) =
  Torus(Point(p1), Point(p2), Point(p3), minor)

```

File: test/primitives.jl [hunk 4]
```diff
@@ -443,5 +443,19 @@
...
+    t = Torus(p₁, p₂, p₃, 1)
+    c = center(t)
+    R, r = radii(t)
+    @test r == 1
+    @test norm(p₁ - c) ≈ R
+    @test norm(p₂ - c) ≈ R
+    @test norm(p₃ - c) ≈ R
+    @test p₁ ∈ t
+    @test p₂ ∈ t
+    @test p₃ ∈ t
```
Comment (juliohm) [inline test/primitives.jl:460]:
```suggestion

    # torus passing through three points
    p₁ = P3(0,0,0)
    p₂ = P3(1,2,3)
    p₃ = P3(3,2,1)
    t = Torus(p₁, p₂, p₃, 1)
    c = center(t)
    R, r = radii(t)
    @test r == 1
    @test norm(p₁ - c) ≈ R
    @test norm(p₂ - c) ≈ R
    @test norm(p₃ - c) ≈ R
    @test p₁ ∈ t
    @test p₂ ∈ t
    @test p₃ ∈ t
    
    # constructor with tuples
    c₁ = T.((0,0,0))
    c₂ = T.((1,2,3))
    c₃ = T.((3,2,1))
    q = Torus(c₁, c₂, c₃, 1)
    @test q == t
```

**Pull request merged**